Modify CustomerStore to support searchCustomers action

iamgabrielma opened the pull request:
Part of https://github.com/woocommerce/woocommerce-ios/issues/7741 , branched from https://github.com/woocommerce/woocommerce-ios/pull/7780 so that should be merged first.

### Description

This PR modifies the existing `CustomerStore` and `CustomerAction` to implement both `CustomerRemote`, and `WCAnalyticsCustomerRemote` as well as their respective actions under the same class.

It also provides a temporary implementation for the `searchCustomers()` action, which returns a `Result<Void, Error>` type to a completion handler, this performs the customer search by name, and is intended to be saved on Core Data in future iterations as "search results".

### Changes:
- Declare a `customerRemote` and a `searchRemote` under `CustomerStore`.
- Implements the `searchCustomers()` action, which performs the customer search by keyword. Only searches by name.
- Modifies the "magnifying glass" button in "Orders" > "+" > "Create new Order" > "+ Add Customer" that calls the action.

### Testing instructions:

1. The current implementation assumes that you have at least one customer in your store, and that customer name contains the word "hello", if that's not the case, you can change this line to the keyword you prefer:

https://github.com/woocommerce/woocommerce-ios/blob/0e33748fcdb9cf46506b6ddef2e6f923a5e6e04a/WooCommerce/Classes/ViewRelated/Orders/Order%20Creation/CustomerSection/CreateOrderAddressFormViewModel.swift#L108

3. Go to "Orders" > "+" > "Create new Order" > "+ "Add Customer" and tap in the "Magnifying glass" icon at the top of the screen

<img width=400 src=https://user-images.githubusercontent.com/3812076/192192292-26a940ff-353f-4c1b-99ba-54d443c0a91d.png>

4. Confirm in Proxyman or any other tool that we send a GET HTTP request to /wc-analytics/customers?search={your-keyword} and that returns a 200 response, example:

```
https://public-api.wordpress.com/rest/v1.1/jetpack-blogs/121710041/rest-api/?json=true&path=/wc-analytics/customers%26_method%3Dget&query=%7B%22search%22%3A%22hello%22%7D
```

```
{
  "data": [
    {
      "id": 99,
      "user_id": 4,
      "username": "hello.merch",
      "name": "hello merch",
      "email": "gabriel.maldonado+testuser20221004@automattic.com",
      "country": "US",
      "city": "soller",
      "state": "CA",
      "postcode": "01101",
      "date_registered": "2022-10-04T08:36:49",
      "date_last_active": "2022-10-04T08:36:49",
      "date_last_order": "2022-10-04 08:36:49",
      "orders_count": 1,
      "total_spend": 1,
      "avg_order_value": 1,
      "date_registered_gmt": "2022-10-04T08:36:49",
      "date_last_active_gmt": "2022-10-04T08:36:49",
      "_links": {
        "customer": [
          {
            "href": "https://mywootestingstore.mystagingwebsite.com/wp-json/wc-analytics/customers/99"
          }
        ],
        "collection": [
          {
            "href": "https://mywootestingstore.mystagingwebsite.com/wp-json/wc-analytics/customers"
          }
        ]
      }
    }
  ]
}
```

iamgabrielma committed:
Register action support for retrieveCustomers

iamgabrielma committed:
Add action to customerSearchTapped

joshheald commented on Yosemite/Yosemite/Actions/WCAnalyticsCustomerAction.swift:
```diff
@@ -0,0 +1,18 @@
+    ///- `onCompletion`: Invoked when the operation finishes.
+    ///     - `result.success(Customer)`: The Customer object
+    ///     - `result.failure(Error)`: Error fetching Customer
+    case retrieveCustomers(
```
`searchCustomers` would be more descriptive of the action

joshheald commented on Networking/Networking/Remote/WCAnalyticsCustomerRemote.swift:
```diff
@@ -9,18 +9,19 @@ public class WCAnalyticsCustomerRemote: Remote {
     ///     - completion: Closure to be executed upon completion.
     ///
-    func retrieveCustomersByName(for siteID: Int64, with name: String, completion: @escaping (Result<[WCAnalyticsCustomer], Error>) -> Void) {
+    public func retrieveCustomersByName(for siteID: Int64, with name: String, completion: @escaping (Result<[WCAnalyticsCustomer], Error>) -> Void) {
```
`searchCustomersByName`

iamgabrielma committed:
Move search Action into CustomerStore

iamgabrielma committed:
Add initial WCAnalyticsCustomer Store and Action

iamgabrielma committed:
Modify JetpackRequest to accept search param

iamgabrielma commented on Yosemite/Yosemite/Model/Model.swift:
```diff
@@ -147,6 +147,7 @@ public typealias PrintingResult = Hardware.PrintingResult
 public typealias CardPresentReceiptParameters = Hardware.CardPresentReceiptParameters
 public typealias CardPresentTransactionDetails = Hardware.CardPresentTransactionDetails
 public typealias StripeAccount = Networking.StripeAccount
+public typealias WCAnalyticsCustomer = Networking.WCAnalyticsCustomer
```
TODO: remove.

iamgabrielma replied:
```suggestion

```

iamgabrielma committed:
Update Yosemite/Yosemite/Model/Model.swift

joshheald reviewed (approved):
Gone through in detail during pairing.

Looks good, once the test fix from the base PR is merged in. :shipit:

iamgabrielma committed:
Merge branch 'trunk' into issue/7741-wcanalyticscustomer-store-actions

iamgabrielma committed:
Merge branch 'issue/7741-wcanalyticscustomer-store-actions' of https://github.com/woocommerce/woocommerce-ios into issue/7741-wcanalyticscustomer-store-actions

iamgabrielma committed:
Improved inline comments

mokagio committed:
Rebundle with Bundler version 2.3.23

This should fix the CI error see in
https://buildkite.com/automattic/woocommerce-ios/builds/8230
